File: amego/invoice_cancel.ts
/**
 * TypeScript version of invoice_cancel.py
 * Invoice Cancellation (作廢發票)
 *
 * This script demonstrates cancelling an invoice using the Amego Invoice API.
 * It matches the Python invoice_cancel implementation exactly.
 *
 * Run with: npx tsx print/invoice_cancel.ts
 */

import * as crypto from "crypto";

// --- Configuration ---
const API_URL = "https://invoice-api.amego.tw/json/f0501";
const APP_KEY = "sHeq7t8G1wiQvhAuIM27";
const INVOICE_TAX_ID = "12345678"; // 統編

// --- Utility: MD5 Hash ---
function md5(data: string): string {
  return crypto.createHash("md5").update(data).digest("hex");
}

// --- Utility: Escape string to match Python's json.dumps (ASCII-only) ---
function escapeStringPython(str: string): string {
  let result = "";
  for (let i = 0; i < str.length; i++) {
    const char = str[i];
    const code = str.charCodeAt(i);

    // Escape special characters
    if (char === '"') result += '\\"';
    else if (char === "\\") result += "\\\\";
    else if (char === "\n") result += "\\n";
    else if (char === "\r") result += "\\r";
    else if (char === "\t") result += "\\t";
    // Python's json.dumps escapes non-ASCII characters to \uXXXX
    else if (code > 127) {
      result += "\\u" + code.toString(16).padStart(4, "0");
    } else {
      result += char;
    }
  }
  return result;
}

// --- Utility: Match Python's json.dumps(indent=0) ---
// Python's json.dumps(obj, indent=0) produces output like:
// '{\n"key": "value",\n"key2": 123\n}'
// Note: newlines between entries, space after colon, no trailing comma
// Also escapes non-ASCII characters to \uXXXX format
function jsonDumpsPythonIndent0(obj: any): string {
  if (Array.isArray(obj)) {
    const arrayItems = obj.map((item) => {
      if (typeof item === "object" && item !== null) {
        return jsonDumpsPythonIndent0(item);
      }
      return typeof item === "string"
        ? `"${escapeStringPython(item)}"`
        : String(item);
    });
    return "[\n" + arrayItems.join(",\n") + "\n]";
  }

  const entries: string[] = [];

  for (const [key, value] of Object.entries(obj)) {
    let valueStr: string;

    if (typeof value === "string") {
      valueStr = `"${escapeStringPython(value)}"`;
    } else if (typeof value === "object" && value !== null) {
      valueStr = jsonDumpsPythonIndent0(value);
    } else {
      valueStr = String(value);
    }

    entries.push(`"${key}": ${valueStr}`);
  }

  return "{\n" + entries.join(",\n") + "\n}";
}

// --- Main Function ---
async function cancelInvoice(invoiceNumber: string): Promise<any> {
  // Business Parameters
  // API expects an Array of objects with CancelInvoiceNumber
  const businessParams = [
    {
      CancelInvoiceNumber: invoiceNumber,
    },
  ];

  // Convert to JSON matching Python's json.dumps(indent=0)
  const apiDataJson = jsonDumpsPythonIndent0(businessParams);

  // Unix Timestamp (10 digits, no milliseconds)
  const currentTimestamp = Math.floor(Date.now() / 1000);

  // Generate MD5 signature
  const hashText = `${apiDataJson}${currentTimestamp}${APP_KEY}`;
  const signature = md5(hashText);

  console.log("=".repeat(60));
  console.log("Invoice Cancellation (作廢發票)");
  console.log("=".repeat(60));
  console.log(`\nAPI URL: ${API_URL}`);
  console.log(`Invoice Number: ${invoiceNumber}`);
  console.log(`Timestamp: ${currentTimestamp}`);
  console.log(`Signature: ${signature}`);

  // Prepare POST data
  const postData = new URLSearchParams();
  postData.append("invoice", INVOICE_TAX_ID);
  postData.append("data", apiDataJson);
  postData.append("time", currentTimestamp.toString());
  postData.append("sign", signature);

  try {
    console.log("\n" + "=".repeat(60));
    console.log("Sending request to API...");
    console.log("=".repeat(60));

    const response = await fetch(API_URL, {
      method: "POST",
      headers: {
        "Content-Type": "application/x-www-form-urlencoded",
      },
      body: postData.toString(),
    });

    if (!response.ok) {
      throw new Error(`HTTP Error: ${response.status} ${response.statusText}`);
    }

    const result = await response.json();

    console.log("\n" + "=".repeat(60));
    console.log("API Response:");
    console.log("=".repeat(60));
    console.log(JSON.stringify(result, null, 2));

    if (result.code === 0) {
      console.log(
        `\n✅ SUCCESS! Invoice ${invoiceNumber} cancelled successfully.`,
      );
    } else {
      console.log(`\n❌ ERROR: ${result.msg || "Unknown error"}`);
    }

    return result;
  } catch (error: any) {
    console.error("\n❌ Request failed:", error.message);
    return null;
  }
}

// --- Run ---
if (require.main === module) {
  const TEST_INVOICE_NUMBER = "AB12345678";
  cancelInvoice(TEST_INVOICE_NUMBER);
}

export { jsonDumpsPythonIndent0, md5, cancelInvoice };

File: amego/invoice_cancel.py
####
# 參考範例 API - 作廢發票
# https://invoice.amego.tw/api_doc/#api-%E7%99%BC%E7%A5%A8-Invoice_Cancel
####

import requests
import json
import time
import hashlib
from typing import Dict, Any, Optional

# Configuration
API_URL = "https://invoice-api.amego.tw/json/f0501"
APP_KEY = "sHeq7t8G1wiQvhAuIM27"
INVOICE_TAX_ID = "12345678"  # 統編


def cancel_invoice(invoice_number: str) -> Optional[Dict[str, Any]]:
    """
    Cancel an invoice using the Amego Invoice API.

    Args:
        invoice_number: The invoice number to cancel (e.g., "AB12345678")

    Returns:
        API response dict if successful, None otherwise.
    """

    # 1. Prepare Business Parameters
    # API expects a list of objects, each with 'CancelInvoiceNumber'
    business_params = [
        {
            "CancelInvoiceNumber": invoice_number
        }
    ]

    # 2. Serialize to JSON string
    # Using indent=0 to match reference implementation requirement
    api_data_json = json.dumps(business_params, indent=0)

    # 3. Generate Request Signature
    current_timestamp = int(time.time())

    # Signature = MD5( api_data_json + timestamp + app_key )
    signature_source = f"{api_data_json}{current_timestamp}{APP_KEY}"
    signature = hashlib.md5(signature_source.encode("utf-8")).hexdigest()

    # 4. Construct Final Request Payload
    request_payload = {
        "invoice": INVOICE_TAX_ID,
        "data": api_data_json,
        "time": current_timestamp,
        "sign": signature
    }

    headers = {
        'Content-Type': 'application/x-www-form-urlencoded'
    }

    print(f"Sending request to: {API_URL}")
    print(f"Invoice Number: {invoice_number}")

    try:
        response = requests.post(
            API_URL, headers=headers, data=request_payload)
        response.raise_for_status()

        resp_json = response.json()

        # Check API business code
        code = resp_json.get('code')
        if code == 0:
            print(
                f"\n[SUCCESS] Invoice {invoice_number} cancelled successfully")
            return resp_json
        else:
            print(f"\n[ERROR] API returned error code: {code}")
            print(f"Message: {resp_json.get('msg')}")
            return resp_json

    except requests.exceptions.RequestException as e:
        print(f"\n[CRITICAL] HTTP Request failed: {e}")
        return None
    except json.JSONDecodeError:
        print(f"\n[CRITICAL] Failed to decode JSON response.")
        print(f"Raw text: {response.text}")
        return None


if __name__ == "__main__":
    # Example usage
    TEST_INVOICE_NUMBER = "AB12345678"
    result = cancel_invoice(TEST_INVOICE_NUMBER)

    if result:
        print(f"\nFull Response:")
        print(json.dumps(result, indent=2, ensure_ascii=False))
